### PR TITLE
gprof: move gprof function from sched to libbuiltin/libgcc

### PR DESCRIPTION
**Pull request by W-M-R**
## Summary
1. Enable interrupt gprof please config CONFIG_PROFILE_MINI
2. Enable instuction gprof please add compile opt "-pg" or config CONFIG_PROFILE_ALL
## Impact
None
## Testing
1. Enable interrupt gprof please config CONFIG_PROFILE_MINI
2. Enable instuction gprof please add compile opt "-pg" or config CONFIG_PROFILE_ALL

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



No, this PR description does not adequately meet the NuttX requirements.  Here's why and how to fix it:

**Problems:**

* **Summary is insufficient:**  It only states *what* was done, not *why*.  What problem does this solve?  What part of the code was changed (files/modules)? How does it work (e.g., what mechanism is used to enable interrupt profiling)?
* **Impact is overly simplistic:**  "None" is unlikely to be entirely true.  At a minimum, enabling profiling impacts performance (overhead).  There might be memory usage implications.  Think critically about *all* potential impacts.
* **Testing is a repeat of the summary:** It doesn't provide any actual test results.  "Testing logs before change" and "Testing logs after change" are empty.  What was the expected outcome?  What was the observed outcome?  Show evidence that the changes work.

**How to improve the PR description:**

**Summary:**

* Explain the motivation for adding gprof support.  Was it requested by a user? Is it for debugging/performance analysis?
* Specify the modified files/modules.  E.g., "Modified the `arch/arm/src/xxx` files and the `sched/` directory to support gprof profiling..."
* Describe the implementation details. E.g., "Interrupt profiling is enabled by setting CONFIG_GPROF_PROFILING, which adds a timer interrupt handler that samples the program counter. Instruction profiling is enabled by adding the -pg compiler option or setting CONFIG_GPROF_ALL..."
* If applicable, link to a related NuttX issue.

**Impact:**

* **Impact on user:**  "Users will need to enable CONFIG_GPROF_PROFILING or compile with -pg to use this feature. Enabling profiling will introduce performance overhead."
* **Impact on build:**  "The build system will now recognize the CONFIG_GPROF_PROFILING option."  (Or if there are other build system changes, mention them).
* **Impact on hardware:** Be specific. Does this work on all architectures? Only certain ones?  "This feature is implemented for ARM and RISC-V architectures." (Or whichever are supported).
* **Impact on documentation:**  "Documentation updates are needed to explain how to enable and use gprof profiling." (And ideally, you would include those documentation updates in the PR).
* **Impact on security, compatibility:**  Consider these carefully.  Even if the answer is NO, explicitly state it.  If there are compatibility concerns (e.g., with older NuttX versions), explain them.


**Testing:**

* **Build Host(s):**  Provide details. "Linux, x86_64, GCC 12.2"
* **Target(s):** Provide specific board and configuration information.  "QEMU-rv32ima, `nsh` configuration."  "STM32F4 Discovery, `hello` configuration."
* **Testing logs before change:**  If applicable, show that profiling *wasn't* working before.  Or state "N/A - Feature did not exist previously."
* **Testing logs after change:** Show the output of `gprof` or other evidence that profiling is now working.  Include relevant snippets of the output, not entire log files. Demonstrate both interrupt and instruction profiling if implemented.


By providing this level of detail, you make it much easier for reviewers to understand and approve your PR.

File: arch/arm/src/cmake/armclang.cmake [hunk 1]
```diff
@@ -120,7 +120,7 @@ if(CONFIG_SCHED_GCOV)
   add_compile_options(-fprofile-generate -ftest-coverage)
 endif()
 
-if(CONFIG_SCHED_GPROF_ALL)
+if(CONFIG_GPROF_ALL)
```
Comment (xiaoxiang781216) [inline arch/arm/src/cmake/armclang.cmake:123]:
```suggestion
if(CONFIG_PROFILE_ALL)
```

File: arch/sim/Kconfig [hunk 2]
```diff
@@ -92,7 +92,7 @@ config SIM_UBSAN_DUMMY
 
 config SIM_GPROF
```
Comment (xiaoxiang781216) [inline arch/sim/Kconfig:93]:
```suggestion
config SIM_PROFILE
```

File: libs/libbuiltin/Kconfig [hunk 3]
```diff
@@ -35,6 +35,30 @@ config COVERAGE_COMPILER_RT
...
+	---help---
+		Enable gprof profiling support.  This will cause the compiler to
+		generate additional code to support profiling.  This will also
+		cause the linker to include the gmon.out file in the final
+		executable.
+		Add the "-pg" parameter to the Makefile when compiling to obtain
+		the function call graph of the specified module. If you do this,
+		please enable "CONFIG_FRAME_POINTER"
+
+config GPROF_ALL
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:51]:
```suggestion
config PROFILE_ALL
```

File: libs/libbuiltin/Kconfig [hunk 4]
```diff
@@ -35,6 +35,30 @@ config COVERAGE_COMPILER_RT
 	select LIB_COMPILER_RT
 	default n
 
+config GPROF_PROFILING
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/Kconfig:38]:
```suggestion
config PROFILE_MINI
```

File: libs/libbuiltin/libgcc/gprof/profile_monitor.c [hunk 5]
```diff
@@ -0,0 +1,584 @@
+/****************************************************************************
+ * libs/libbuiltin/libgcc/gprof/profile_monitor.c
```
Comment (xiaoxiang781216) [inline libs/libbuiltin/libgcc/gprof/profile_monitor.c:2]:
```suggestion
 * libs/libbuiltin/libgcc/profile.c
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**